Parquet: Update parquet to 1.13.1

singhpk234 opened the pull request:
### About the change

parquet 1.13.0 release notes : https://github.com/apache/parquet-mr/blob/apache-parquet-1.13.0/CHANGES.md?plain=1#L22-L78

This has fix for : https://issues.apache.org/jira/browse/PARQUET-2160

earlier iceberg pr to fix issue observed above : 
- https://github.com/apache/iceberg/pull/5681 

cc @rdblue @bryanck

bryanck commented:
Should we also revert https://github.com/apache/iceberg/pull/5681 ?

singhpk234 commented:
> Should we also revert https://github.com/apache/iceberg/pull/5681 ?

IMHO we should, was thinking of the same but in a separate commit, post validating the issues. 
Presently this upgrade takes a dependency on hadoop 3.2, and interestingly some of the BloomFilter ut's fails as well. Taking a look into the same as well.

Fokko commented:
I've created a PR to support Hadoop 2.7.3 as well: https://github.com/apache/parquet-mr/pull/1084

Fokko commented:
Looks like this commit made the bloom tests on the Iceberg side fail:
```
➜  parquet-mr git:(4e9e79c89) ✗  git bisect bad
4e9e79c895e61775066e11240729b574e2264b8b is the first bad commit
commit 4e9e79c895e61775066e11240729b574e2264b8b
Author: ChenLiang.Lu <31469905+yabola@users.noreply.github.com>
Date:   Mon Feb 27 15:45:46 2023 +0800

    PARQUET-2251 Avoid generating Bloomfilter when all pages of a column are encoded by dictionary in parquet v1 (#1033)

 .../apache/parquet/hadoop/ParquetFileWriter.java   |   3 +-
 .../apache/parquet/hadoop/TestBloomFiltering.java  |  18 ++-
 .../parquet/hadoop/TestStoreBloomFilter.java       | 132 +++++++++++++++++++++
 3 files changed, 151 insertions(+), 2 deletions(-)
 create mode 100644 parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestStoreBloomFilter.java
```

singhpk234 commented:
> Looks like this commit made the bloom tests on the Iceberg side fail:

This is an interesting find @Fokko, this implies, even when we are enabling BF via iceberg table conf's (https://iceberg.apache.org/docs/latest/configuration/#write-properties) parquet may decide not to write a BF for the column if all it's pages are dictionary encoded (I am assuming this is because there will be no benefit of writing BF in this particular case), Do we need to mention this in our doc as well, as the given table prop is a hint for parquet to write BF, but it may not choose to ?

Fokko commented:
@singhpk234 Yes, that's indeed the case. The dictionary is a better version of the bloom filter because it cannot contain false positives, this makes it redundant to also generate the bloom filter.

singhpk234 commented:
@Fokko, added a fix for the failing UTs dues to BF changes and also updated the docs with the current behaviour of parquet, I think once your pr is in https://github.com/apache/parquet-mr/pull/1084 and parquet 1.13.1 is out we should be good to go, please let me know your thoughts !

Fokko commented:
Ran a benchmark on the 1.2.1 branch:
```
Apache Iceberg 1.2.1 with Apache Iceberg 1.12.3

Benchmark                                                                            Mode  Cnt  Score   Error  Units
IcebergSourceFlatParquetDataReadBenchmark.readFileSourceNonVectorized                  ss    5  5,068 ± 0,151   s/op
IcebergSourceFlatParquetDataReadBenchmark.readFileSourceVectorized                     ss    5  1,961 ± 0,081   s/op
IcebergSourceFlatParquetDataReadBenchmark.readIceberg                                  ss    5  2,438 ± 0,028   s/op
IcebergSourceFlatParquetDataReadBenchmark.readWithProjectionFileSourceNonVectorized    ss    5  1,061 ± 0,120   s/op
IcebergSourceFlatParquetDataReadBenchmark.readWithProjectionFileSourceVectorized       ss    5  0,490 ± 0,186   s/op
IcebergSourceFlatParquetDataReadBenchmark.readWithProjectionIceberg                    ss    5  0,377 ± 0,052   s/op

Apache Iceberg 1.2.1 with Apache Parquet 1.13.1-SNAPSHOT

Benchmark                                                                            Mode  Cnt  Score   Error  Units
IcebergSourceFlatParquetDataReadBenchmark.readFileSourceNonVectorized                  ss    5  4,509 ± 0,047   s/op
IcebergSourceFlatParquetDataReadBenchmark.readFileSourceVectorized                     ss    5  2,137 ± 0,176   s/op
IcebergSourceFlatParquetDataReadBenchmark.readIceberg                                  ss    5  2,446 ± 0,056   s/op
IcebergSourceFlatParquetDataReadBenchmark.readWithProjectionFileSourceNonVectorized    ss    5  1,033 ± 0,085   s/op
IcebergSourceFlatParquetDataReadBenchmark.readWithProjectionFileSourceVectorized       ss    5  0,471 ± 0,043   s/op
IcebergSourceFlatParquetDataReadBenchmark.readWithProjectionIceberg                    ss    5  0,372 ± 0,039   s/op
```

Fokko commented:
@singhpk234 can I invite you to give the 1.13.1 RC a try: https://lists.apache.org/thread/0yokbmfcbhz76ftjbxktwxfo5vrt57od Would be great if you can vote on the RC.

singhpk234 commented:
Thank you @Fokko for the awesome work ! Added my +1 :) !

singhpk234 commented on build.gradle:
```diff
@@ -137,6 +137,12 @@ allprojects {
   repositories {
     mavenCentral()
     mavenLocal()
+    // testing purpose only, to be removed when spark-3.3 rc passes
+    maven {
+      url "https://repository.apache.org/content/groups/staging/"
+      mavenContent {
+      }
+    }
```
will remove when 1.13.1 is officially released :) !

singhpk234 commented on build.gradle:
```diff
@@ -137,6 +137,11 @@ allprojects {
   repositories {
     mavenCentral()
     mavenLocal()
+    maven {
+      url "https://repository.apache.org/content/groups/staging/"
+      mavenContent {
+      }
+    }
```
will remove when parquet 1.13.1 is officially released

Fokko replied:
1.13.1 is out: https://parquet.apache.org/blog/2023/05/18/1.13.1/

🥳

Fokko commented on aliyun/src/test/java/org/apache/iceberg/aliyun/oss/mock/AliyunOSSMockLocalStore.java:
```diff
@@ -36,8 +36,8 @@
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import org.apache.commons.codec.binary.Hex;
```
Looking at https://github.com/apache/iceberg/pull/5024/files#r896336657 We'd rather avoid using `apache.commons`.

Fokko replied:
I noticed that it all over the codebase, I've created a separate ticket: https://github.com/apache/iceberg/issues/7658

amogh-jahagirdar commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -339,6 +344,8 @@ private static class Context {
       private final int rowGroupSize;
       private final int pageSize;
       private final int pageRowLimit;
+
```
Nit: do we need the newline here?

amogh-jahagirdar commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -339,6 +344,8 @@ private static class Context {
       private final int rowGroupSize;
       private final int pageSize;
       private final int pageRowLimit;
+
```
Nit: don't think we need the newline here?

singhpk234 replied:
ACK

amogh-jahagirdar reviewed (approved):
Non-blocking nit, this looks great @singhpk234 !

Fokko commented:
Thanks @singhpk234 for picking this up, and @amogh-jahagirdar for the review!

@bryanck do you want the honor of reverting https://github.com/apache/iceberg/pull/5681?

rdblue commented on docs/configuration.md:
```diff
@@ -47,51 +47,52 @@ Iceberg tables support table properties to configure table behavior, like the de
+| write.parquet.row-group-size-bytes                   | 134217728 (128 MB)          | Parquet row group size                                                                                                                                                                            |
+| write.parquet.page-size-bytes                        | 1048576 (1 MB)              | Parquet page size                                                                                                                                                                                 |
+| write.parquet.page-row-limit                         | 20000                       | Parquet page row limit                                                                                                                                                                            |
+ | write.parquet.dictionary.enabled                     | true                        | Enable dictionary encoding                                                                                                                                                                        |
```
This doesn't match the setting above.

rdblue commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -135,6 +135,9 @@ private TableProperties() {}
   public static final String DELETE_PARQUET_PAGE_ROW_LIMIT = "write.delete.parquet.page-row-limit";
   public static final int PARQUET_PAGE_ROW_LIMIT_DEFAULT = 20_000;
 
+  public static final String PARQUET_DICT_ENABLED = "write.parquet.enable.dictionary";
```
This doesn't match the format of Iceberg options. It should be `write.parquet.dict-enabled` to match `dict-size-bytes` or should use `enabled` as the last word if you're using `dictionary` as a part of the hierarchy.

aokolnychyi replied:
+1, the name does match our pattern

singhpk234 replied:
Apologies for my oversight here, i intended to have `.enabled` in suffix which i mentioned in the table prop but messed up here.

rdblue commented on docs/configuration.md:
```diff
@@ -47,51 +47,52 @@ Iceberg tables support table properties to configure table behavior, like the de
+| write.update.mode                                    | copy-on-write               | Mode used for update commands: copy-on-write or merge-on-read (v2 only)                                                                                                                           |
+| write.update.isolation-level                         | serializable                | Isolation level for update commands: serializable or snapshot                                                                                                                                     |
+| write.merge.mode                                     | copy-on-write               | Mode used for merge commands: copy-on-write or merge-on-read (v2 only)                                                                                                                            |
+| write.merge.isolation-level                          | serializable                | Isolation level for merge commands: serializable or snapshot                                                                                                                                      |
```
Why was this entire table reformatted?

rdblue commented on parquet/src/test/java/org/apache/iceberg/parquet/TestBloomRowGroupFilter.java:
```diff
@@ -197,6 +198,7 @@ public void createInputFile() throws IOException {
     try (FileAppender<Record> appender =
         Parquet.write(outFile)
             .schema(FILE_SCHEMA)
+            .set(PARQUET_DICT_ENABLED, "false")
```
@singhpk234 @Fokko, I don't think it makes sense to create an option that is public and must be supported by Iceberg moving forward just for this test case. Is it possible to set this with a Hadoop option or to do this some other way?

bryanck commented:
> Thanks @singhpk234 for picking this up, and @amogh-jahagirdar for the review!
> 
> @bryanck do you want the honor of reverting #5681?

Sure, PR incoming…

amogh-jahagirdar commented:
I guess there were a few oversights here that we want to address before release, I was thinking that it would make sense for certain use cases to disable dictionary encoding if they wanted to use bloom filters (controlling the space tradeoff). 

So we should conclude

1.) Do we really want a table property? Curious to know other's opinions. cc: @Fokko @rdblue @singhpk234 @aokolnychyi 

2.) If the answer is yes, let's go back and correct the naming in the properties.

aokolnychyi commented:
I agree with @amogh-jahagirdar. It is probably not worth adding an extra table property for the sake of the test. I doubt anyone would ever configure it. Can we fix the test?

bryanck commented:
I opened https://github.com/apache/iceberg/pull/7664 to revert the original workaround that is no longer needed with the upgrade to Parquet 1.13.
